framework/src/constraints/ADMortarLagrangeConstraint.C

```diff
@@ -0,0 +1,177 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "ADMortarLagrangeConstraint.h"
+
+// MOOSE includes
+#include "MooseVariable.h"
+#include "Assembly.h"
+#include "SystemBase.h"
+#include "ADUtils.h"
+
+InputParameters
+ADMortarLagrangeConstraint::validParams()
+{
+  InputParameters params = ADMortarConstraint::validParams();
+  return params;
+}
+
+ADMortarLagrangeConstraint::ADMortarLagrangeConstraint(const InputParameters & parameters)
+  : ADMortarConstraint(parameters)
+{
+}
+
+void
+ADMortarLagrangeConstraint::computeJacobian(Moose::MortarType mortar_type)
+{
+  std::vector<DualReal> residuals;
+  size_t test_space_size = 0;
+  typedef Moose::ConstraintJacobianType JType;
+  typedef Moose::MortarType MType;
+  std::vector<JType> jacobian_types;
+  std::vector<dof_id_type> dof_indices;
+
+  switch (mortar_type)
+  {
+    case MType::Secondary:
+      dof_indices = _secondary_var.dofIndices();
+      jacobian_types = {JType::SecondarySecondary, JType::SecondaryPrimary, JType::SecondaryLower};
+      break;
+
+    case MType::Primary:
+      dof_indices = _primary_var.dofIndicesNeighbor();
+      jacobian_types = {JType::PrimarySecondary, JType::PrimaryPrimary, JType::PrimaryLower};
+      break;
+
+    case MType::Lower:
+      if (_var)
+        dof_indices = _var->dofIndicesLower();
+      jacobian_types = {JType::LowerSecondary, JType::LowerPrimary, JType::LowerLower};
+      break;
+  }
+  test_space_size = dof_indices.size();
+
+  residuals.resize(test_space_size, 0);
+  for (_qp = 0; _qp < _qrule_msm->n_points(); _qp++)
+    for (_i = 0; _i < test_space_size; _i++)
+      residuals[_i] += _JxW_msm[_qp] * _coord[_qp] * computeQpResidual(mortar_type);
```

Comment:
I think we would like to make life in user code as easy as possible. For instance I see in `NormalMortarMechanicalContact` that you have high-level 0 residual that you return if `_i` is not on the face. I would instead propose a design in which you only call `computeQpResidual` when `_i` is on the face. So I would move this code after your dof filtering, and in your dof filtering I would build up something like a `std::vector<unsigned int> is` which only contains the `_i`s that are on the face. Then you can do
```c++
  for (const auto index : is)
  {
    _i = index;
    // call computeQpResidual and stuff
  }
```
I recognize that this then will require you to implement `computeResidual` as well as `computeJacobian` in this class

Reply:
Then user code should essentially never be returning 0 from `computeQpResidual`

Reply:
Can you please inspect the code to see if that's what you described.

Reply:
Never mind, after latest commit going back to:

```Assertion `compare_dofs_set == current_dofs_set' failed
We're going to see whether the dof sets are the same. IIRC the degree of freedom dependence (as indicated by the dof index set held by the ADReal) has to be the same for every residual passed to this method otherwise constrain_element_matrix will not work.
at /Users/marta3/projects3/moose/framework/build/header_symlinks/Assembly.h, line 2798```